worldbank/DECAT_Space2Stats
# Feature/cli ingest

## What I Changed

- Implemented a CLI tool for ingestion of Parquet files into PostgreSQL using Typer.
- Added commands for downloading a Parquet file from S3, loading it into the database, and a combined `download-and-load` command.
- Updated the ingestion process to use chunked loading with a progress bar.
- Modified the database acceptance test to reflect CLI-based ingestion.

## How I Changed It

- Created a new Python module `space2stats_ingest` with the CLI tool.
- Added three commands:
  - `download`: Downloads a Parquet file from an S3 bucket.
  - `load`: Loads a local Parquet file into a PostgreSQL database.
  - `download-and-load`: Downloads a Parquet file from S3 and loads it into the database.
- Updated the Pyproject.toml to include the CLI tool under `poetry.scripts` for easier accessibility.
- Enhanced data loading to use chunked processing for improved performance during ingestion.

## How to Test It

1. Set up your AWS environment variables:
   ```bash
   export AWS_ACCESS_KEY_ID=your_access_key
   export AWS_SECRET_ACCESS_KEY=your_secret_key
   export AWS_REGION=your_region
   ```

2. Spin up the database using `docker-compose` (if necessary):
   ```bash
   docker-compose up
   ```

3. Install the necessary dependencies:
   ```bash
   poetry install
   ```

4. Use the CLI to download and load a Parquet file:
   ```bash
   poetry run space2stats-ingest download-and-load "s3://yourbucket/space2stats_updated.parquet" "postgresql://postgres:password@localhost:5432/postgis"
   ```

5. Run the `download` and `load` commands individually:
   ```bash
   poetry run space2stats-ingest download "s3://yourbucket/space2stats_updated.parquet" --local-path "local.parquet"
   poetry run space2stats-ingest load "postgresql://postgres:password@localhost:5432/postgis" --parquet-file "local.parquet"
   ```

6. Verify that the data is loaded into the `space2stats` table in PostgreSQL using SQL queries.

## Other Notes

- Added chunked loading with a progress bar to optimize ingestion of large datasets.
- The `space2stats_ingest` CLI is now accessible on the user's `PATH` when installed via Poetry.
- Updated the database acceptance test documentation to reflect the new CLI ingestion process.
- Closes #64; @bpstewar to be confirmed.
